### ISSUE TITLE
docker build should support privileged operations

### ISSUE THREAD

**Issue by darklajid** [moby/moby#1916]
Currently there seems to be no way to run privileged operations outside of docker run -privileged. 

That means that I cannot do the same things in a Dockerfile. My recent issue: I'd like to run fuse (for encfs) inside of a container. Installing fuse is already a mess with hacks and ugly workarounds (see [1] and [2]), because mknod fails/isn't supported without a privileged build step.

The only workaround right now is to do the installation manually, using run -privileged, and creating a new 'fuse base image'. Which means that I cannot describe the whole container, from an official base image to finish, in a single Dockerfile.

I'd therefor suggest adding either
- a docker build -privileged  
  this should do the same thing as run -privileged, i.e. removing all caps limitations

or
- a RUNP command in the Dockerfile  
  this should .. well .. RUN, but with _P_rivileges

I tried looking at the source, but I'm useless with go and couldn't find a decent entrypoint to attach a proof of concept, unfortunately. :(

1: https://github.com/rogaha/docker-desktop/blob/master/Dockerfile#L40
2: https://github.com/dotcloud/docker/issues/514#issuecomment-22101217

**Comment by kkannan1982** [comment]
+1 This makes the reproducible docker container at any given point in time( just need to carry the dockerfile alone)

**Comment by ljsommer** [comment]
+1 having to completely tear apart my ansible baseline roles just to work around stuff like this. I was really hoping that docker adoption would let me use a lot of my existing ansible code but not only have I already created custom roles for size but now I am having to work around issues like this.

**Comment by cpuguy83** [comment]
@lsjommer what kinda of things do you have to work around? --privileged is a completely insecure way to run a container and gives the user in the container full root access to the host.

I'm not saying let's not implement this, but let's get real about what we are talking about.

Also this would be relatively easy to implement if someone wants to go for it...

**Comment by ljsommer** [comment]
@cpuguy83 This is from our standard "bare metal" baseline role that I am trying to adopt into docker containers to get all libs installed, and it deals with shared memory but maybe I don't even need to bother with it on the container build and need only run it on the container host?
http://pastebin.com/P3QQxjNQ

I admit that I don't fully understand how Docker handles resource sharing.

**Comment by cpuguy83** [comment]
@ljsommer So setting shm is a different beast all-together and would not persist between RUN commands (or for when you actually `docker run`) anyway.

**Comment by ljsommer** [comment]
@cpuguy83 Yeah I think this was mostly my fault in the regard that I hit what I thought to be an issue that I can just shift onto the baseline for the container host itself.
Thanks for taking the time to respond and apologies for not appropriately educating myself before complaining.
;)

**Comment by malcm** [comment]
Any idea on RUNP / -privileged during build process?
It would be great fo setting ip tables to limit docker access to spcified ip-addresses

**Comment by sakurai-youhei** [comment]
I also want RUNP and/or "docker build --privileged".
```
FROM ubuntu:latest
MAINTAINER xyz

RUN apt-get -qq update
RUN apt-get -yq install iptables

RUN iptables -t nat -I OUTPUT -p tcp --dport 443 -j DNAT --to-destination 127.0.0.1:8080 && iptables-save > /etc/iptables.rules
```
This Dockerfile doesn't work due to following error but works if doing with "docker run --privileged"...
```
getsockopt failed strangely: Operation not permitted
```

**Comment by jpetazzo** [comment]
@malcm, @sakurai-youhei: even if you had something like `RUNP`, it would not work in this scenario, because iptables rules are not persisted in the filesystem.

Let me explain: when you do `RUN x`, Docker executes `x` then takes a snapshot of the filesystem. Things outside of the filesystem (running processes, routing tables, iptables rules, sysctl settings...) are not stored in Docker images.

If you want custom iptables rules, one method is:
- start your container e.g. with `--name myapp`
- start another container, privileged, one-shot, to set up the iptables rule, e.g. `docker run --net container:myapp --privileged iptablesimage iptables -t nat ...`

Does that make sense?

**Comment by sakurai-youhei** [comment]
@jpetazzo: Thank you for your response. In my case, I put second command to make iptables rule persist as data on file system as below. This should enable me to load iptables rule after starting container with --privileged option.
```
RUN do-something-with-iptables && iptables-save > /etc/iptables.rules
```
Without RUNP nor "build --privileged", I'm enforced to write like:
```
ADD iptables.rules /etc/
```
Yes, this might be sufficient, however, I need to add iptables.rules beside Dockerfile in my repo.

That's why I want (or would like to have gently) RUNP. :)

**Comment by dubglan** [comment]
@jpetazzo @strib
Beyond iptables issues, mounts and other privileged operations, I think there's one build scenario we should be addressing.

We ship appliances for deployment in VMs and bare-metal installation. However, for testing we use container environments. In turn, inside of those appliances we run containers. So test containers have to be based on docker-in-docker. Now imagine that we have a service image that needs to be pre-loaded in the test image (so that service images are not downloaded from the registry in test time). Right now we can't do that since we can't run d-in-d container in privileged mode during the build of the Dockerfile that uses d-in-d as a base: docker daemon won't start, "docker pull" or "docker load" won't work.

**Comment by elyscape** [comment]
I had an issue where, when running on a RHEL7 host, `su` would fail if the current user was root.  Weirdly, if the current user was something else, `su` worked fine.  Regardless, the workaround was to pass the run command the flag `--add-cap=SYS_RESOURCE`; due to this issue, however, it wasn't possible to do this during the build step.

**Comment by PonderingGrower** [comment]
+1 Scripting around Dockerfiles with ```docker run``` and ```docker commit`` is ridiculous. Please include this functionality.

**Comment by davidl-zend** [comment]
+1 on the necessity of this feature. I think that a global "security level" could be configured in a config file which limits the capabilities that can be given to a container. There should be a safe default (like today) and a sysadmin could change it to allow containers to run with more privileges. dockerfiles with such a RUNP instruction could fail to run with a message such as "this Dockerfile requires the following capabilities .... to build" on a system which has such global limits.  

I think this allows a balance between security and usability.

**Comment by jave** [comment]
We also have this issue while trying to build an image with an evli proprietary db,which shall remain nameless, inside it.
The db wants to allocate massive amounts of memory, which its not allowed to by docker.

Our current workaround is a 2 phase build with a run --privileged step, and separate commit step.

Maybe we could configure docker to allow the memory allocation some other way. Its a bit hard to find out what the db actually wants to do because its proprietary.

**Comment by zrml** [comment]
+1 
for this feature.
for historic and use case example see this dupe
https://github.com/docker/docker/issues/12138#issuecomment-90536998
thanks @cpuguy83 for pointing out the dupe

**Comment by binhex** [comment]
i too have this issue, trying to hook into a cifs share during docker build is not permitted unless the privileged flag is supplied, any way around this?

**Comment by thaJeztah** [comment]
There's now a pull request that implements this; you can check progress there; https://github.com/docker/docker/issues/12261

**Comment by cpuguy83** [comment]
If something does require privileged mode then it is possibly modifying the host in some way, which means the image may not be portable since these modifications would need to be run on other hosts trying to consume the image.

Once #13171 is merged I think we should close this as it will make rolling your own builder trivial, and as such allowing --privileged.
I don't think the built-in `docker build` should allow this.

**Comment by PerilousApricot** [comment]
So @cpuguy83, if I understand right, the way to support this issue would be to completely reimplement ```docker build``` but with an extra parameter?

**Comment by PerilousApricot** [comment]
I guess to say, once the other patch is pushed through, I'd need to put together my own version of ```docker build``` (maybe ```docker pbuild```?) to fill in the additional functionality?

**Comment by HowardMei** [comment]
Is there any progress on this issue? I checked the PRs mentioned above and all of them failed.  
Is it possible to make a `BUILD --privileged/--granted` option more granular and make granted accesses to a specific group of host resources limited only to the image builder/owner?

**Comment by pospispa** [comment]
+1 for any solution that allows me to do `RUN docker pull` in a Dockerfile.

Use case: I need a bunch of tools for image conversion and documentation build, however, all these tools can't be installed into a single image because of conflicting libraries. That's why I separate some of these tools into a separate image and I would like to distribute all tools in a single image, i.e. image in an image. That's why I want to do `RUN docker pull` in my Dockerfile.

**Comment by vincentwoo** [comment]
@cpuguy83 it doesn't look like this issue was resolved to anyone's satisfaction. I absolutely, 100%, need to be able to do something as dull as write to `/proc/sys/kernel/core_pattern` during a build.

In the current world, I can do this privileged operation via a `run` workaround and just push that image to the hub anyway. Additionally, no `Dockerfile` I have *ever* produced is strictly reproducible, because they pull from random constantly changing public repos. I had no idea that

1. Public consumption of my images was a priority.
2. They had any need, ever, to be reproducible.

People are *going* to do shitty workarounds to get `privileged` in build. I think you should definitely go where your users are and allow this in the core build tool. Add a terrifying message if necessary.

cc @thaJeztah, who seems sympathetic to this position

**Comment by cpuguy83** [comment]
Look, I created a PR to enable this and it was rejected.
I don't see this happening with the builder in any form.

**Comment by vincentwoo** [comment]
It seemed like you made the final call. I'll agitate in the PR thread itself, then.

**Comment by sedwards** [comment]
This is needed for installing the older JDK 1.6 packages under CentOS as its RPMs attempt to register is binmft_misc which fails without running under --privileged.

Dockerbuild is a non-stater for building containers with it.

To replicate

FROM Centos5.11
RUN yum intall -y jre-1.6.0_29-fcs

**Comment by shubhamrajvanshi** [comment]
We need to have privileged command part of build as an optional flag. I am trying to port one of our application to docker as a POC and it fails for one of our component as it has IPtables settings that are not being applied and the build fails. I can do the necessary changes manually and commit it but then what's the fun of docker build. It's just a part of the build and should be easy to port as it already a part of the main release. 

Docker should be easily able to run intermediate containers with privileged option set.

**Comment by thaJeztah** [comment]
@shubhamrajvanshi we're in the process of moving the "builder" out of the daemon (and to the client), this should open the door for more customizations to the build process (including being able to implement custom builders). Possibly allowing "privileged" builds can be considered, but that's a decision that can be made after the refactoring. (See https://github.com/docker/docker/blob/master/ROADMAP.md#122-builder)

**Comment by cpuguy83** [comment]
@shubhamrajvanshi you can't make change to iptables in build for a good reason, the settings would never stick.

There are very few things one can do with --privileged that even makes sense in build.

**Comment by shubhamrajvanshi** [comment]
@thaJeztah Thanks that would be helpful.
@cpuguy83 Would that be the case even if I am using iptables-persistent package on the image.

**Comment by cpuguy83** [comment]
This would save the rules to disk, then they'd still have to be reloaded, unfortunately.

**Comment by karelstriegel** [comment]
+1

**Comment by tlbtlbtlb** [comment]
I would really like this too, to allow using nVidia's CUDA drivers from with Docker on CoreOS. The installer they provide builds a kernel module against the kernel source, and then installs it using modprobe. I can't see how to make it work without some kind of --privileged option to build.

**Comment by cornelf** [comment]
Why not by default always support privileged mode on build?

**Comment by kujiy** [comment]
+1 
I want to use mysql command in Dockerfile for centos7.
Of course we can use entrypoint.sh but it's more useful if we can use -privileged for both build and run.

**Comment by cpuguy83** [comment]
There is no need for --privileged to run the MySQL command.

This issue should be closed as it doesn't seem like this is going to happen (or even should happen).
Allowing privileged in build means allowing the builder to change things on the host which in turn makes it so that image only works on that host (or hosts that have had similar modifications).

**Comment by brunoqc** [comment]
> Allowing privileged in build means allowing the builder to change things on the host which in turn makes it so that image only works on that host (or hosts that have had similar modifications).

Does this apply to the chroot user case?

**Comment by stuaxo** [comment]
I'm trying to work out how to do ```dpkg-depcheck -d ./configure``` without something like this.

During build (or run without --priviledged) I get the error below - I've got no idea how to work out what permission it needs or how to enable it.

```
dpkg-depcheck -d ./configure
strace: test_ptrace_setoptions_followfork: PTRACE_TRACEME doesn't work: Permission denied
strace: test_ptrace_setoptions_followfork: unexpected exit status 1
Running strace failed (command line:
strace -e trace=open,execve -f -q -o /tmp/depchJNii2o ./configure
devel@98013910108c:~/src/cairo-1.14.2$ 
```

**Comment by ccjon** [comment]
After about 3 years and 162 comments, I think it's got enuf interest to do. The comments about a privileged mode not being necessary for most of the cited cases are true, even my own; but should not be used to proscribe against what might be useful for local, temporary, exploratory, and/or expedient builds. Publish warnings until the cows fart out a harmony, print out command line warnings, revile and denounce it's use, but give people flexibility. Portability is not always everyone's primary interest.

**Comment by robeferre** [comment]
+1

**Comment by robeferre** [comment]
+1

I really need to mount a NFS volume inside a docker container, until now I couldn't create the NFS share without the "--privileged=true" flag. The better case in my opinion is to build the image using privileged command.

**Comment by scue** [comment]
+1

```
Step 19 : RUN lxc-create -t ubuntu.sf -n percise -- -r precise -a i386 -b root
 ---> Running in 4c51b7cf0058
lxc_container: lxccontainer.c: create_run_template: 893 error unsharing mounts
lxc_container: lxccontainer.c: create_run_template: 1084 container creation template for percise failed
lxc_container: lxc_create.c: main: 274 Error creating container percise
The command '/bin/sh -c lxc-create -t ubuntu.sf -n percise -- -r precise -a i386 -b root' returned a non-zero code: 1

```

**Comment by eNcacz** [comment]
I'm trying to install gobject-introspection in gentoo system in docker during build but it fails with this error:

 * ISE:_do_ptrace: ptrace(PTRACE_TRACEME, ..., 0x0000000000000000, 0x0000000000000000): Operation not permitted

The same result when I try to install it manually in container but when try it from container launched in privileged mode (docker run --privileged) then it works well.

The same problem when I try to install glibc.

So I also need a way how to run privileged command during build.

I'm using docker version 1.10.1 and the problem with glibc doesn't appears in 1.9

**Comment by wrouesnel** [comment]
We too would like a way to do a privileged docker build. The biggest hurdle I can see from the code is that it would necessitate either editing the `ContainerConfig` struct to store "privileged" as an operational mode for a container (not great - can hide config) or changing the interface for the builder.Backend so `ContainerCreate` can be passed either a `privileged` option or an ExecConfig.

I suspect the ExecConfig would make the most sense.

**Comment by alexey-vostrikov** [comment]
In version 1.10 something broken an we can not build 32 containers
--privileged or --security-opt seccomp:unconfined for BUILD - is really necessary

**Comment by zerthimon** [comment]
big +1 from me

It's a real problem for me that I cannot use 'mount' command during the build.
I was trying to overcome the limitaion of a host directory cannot be mounted into a container during the build, so I set up an NFS server on host and tried mounting an NFS share, just to find out that's it's not possible because of unprivileged mode.

In my use case I need to install some stuff in the image without copying it to the build context, and ADDing it before installing it.

Feels like I'm left without options.

**Comment by alexey-vostrikov** [comment]
> thaJeztah referenced this issue on 10 Mar
> Regression in LTTng behavior after upgrading to 1.10.2 #20818 Closed
No it is not closed, we are using 1.11.0-0~wily and in 32bit cintainers neworking is not working since 1.10.0, but 1.9.x are working great

**Comment by ctindel** [comment]
I am amazed that something so obviously needed by so many people has not been implemented even though people have been begging for it for 2.5 years in this thread alone and given that people have submitted PRs to implement this functionality.

**Comment by PonderingGrower** [comment]
Agreed @ctindel, this issue is one of the reasons why I'm migrating from docker to [rkt](https://github.com/coreos/rkt).

**Comment by cpuguy83** [comment]
@ctindel It's something we're not ready to implement or support. The implementation itself is rather simple (I even implemented it myself so we could discuss), that's not the issue.

`--privileged` is a tank, and allowing it on build is dangerous, and greatly affects image portability.

**Comment by shubhamrajvanshi** [comment]
Brian,

If you have a work around can you please share it with me as well. I would
appreciate that.

Thanks
Shubham Rajvanshi
669-300-8346

**Comment by tlbtlbtlb** [comment]
I don't understand the effect on portability. How do privileged operations at build time affect portability? I mean, it's not hard to create non-portable images in various ways with or without privilege, but is there some way that images built with privileged operations are necessarily non-portable?

**Comment by davidl-zend** [comment]
I don't believe every container must be portable. Some containers are created to share with the community and some might be created for deploying internal applications.

The portability problem with an app that requires running in privileged mode lies in the app itself.

Privileged mode is the means of last resort for getting them the app to work without code changes.

I believe that an image builder who requires privileged mode building understands that such container might also require running in privileged mode.

It should be clearly documented that building in privileged mode is discouraged since it might create portability problems.

Sent from Outlook Mobile<https://aka.ms/blhgte>



On Mon, May 2, 2016 at 2:53 PM -0700, "Trevor Blackwell" <notifications@github.com<mailto:notifications@github.com>> wrote:


I don't understand the effect on portability. How do privileged operations at build time affect portability? I mean, it's not hard to create non-portable images in various ways with or without privilege, but is there some way that images built with privileged operations are necessarily non-portable?

-

**Comment by cpuguy83** [comment]
@tlbtlbtlb Because privileged mode gives you full access to the host. Maybe I set something simple like shmmax or something much worse.
I guarantee these things will happen on day 1 that this is available.

@davidl-zend "portable" doesn't mean sharing with the community. It means moving from one machine to another.

**Comment by ctindel** [comment]
@cpuguy83 As others have pointed out, there are many many other ways to break image portability as well.  All you're doing by not having a process for privileged building is forcing people to have a two step process, either by partially building from a Dockerfile and then manually changing the container and doing a commit, or by partially building from a Dockerfile and finishing the privileged installation the first time the container is launched which sucks because that means if you're doing something time consuming it could take minutes for the first boot to work.

Given the comments I've seen in various forums I'm willing to bet that A LOT of people are doing exactly this already to work around the docker limitation that exists today.  

Once you have an architecture where image portability is broken in dozens of other ways, what point is there in tilting against this specific windmill?

**Comment by cpuguy83** [comment]
@ctindel I would love to see some examples of image portability being broken

Doing this kind of stuff at first container start is _exactly_ the right way to go.
It's a runtime configuration it shouldn't be in the image.

**Comment by ctindel** [comment]
@cpuguy83 Are you debating the fact that someone could do a partial build from a Dockerfile, spin up a container in privileged mode, make more changes, then do a docker commit?  How is that any different than doing a build in privileged mode?  Because that's what other people are doing today.

I am not trying to be testy, I'm just saying this is a severe limitation on the platform that people are working around in an awkward way.

For example, there are 3rd party debian packages (and probably RPMs) that can't be installed correctly unless you have certain capabilities in the system.  Installing a debian package isn't a "runtime configuration" it's a freaking part of the installation process.

**Comment by cpuguy83** [comment]
@ctindel I'm not debating this at all. The difference is supporting the behavior. If there was no difference, we wouldn't be having this discussion.

**Comment by PerilousApricot** [comment]
For me, I'm a consenting adult and want to be able to roll a [packstack](https://wiki.openstack.org/wiki/Packstack) image across a bunch of nodes. Building with a Dockerfile bombs currently, so I have to cheat to get around the docker restrictions.

**Comment by ctindel** [comment]
@cpuguy83 It is still very unclear to me (and I think others in this thread) what exactly is gained by not providing this option that people are working around in other ways.  The sort of architectural purity (I don't have a better word for it) you seem to be arguing for was gone the minute the commit option was added anyway, so I really do not understand what difference there is whether it is done via a Dockerfile with privileged build or via a docker commit in a privileged container. 

Except that one way is a freaking PITA for people and one way slots into the current mechanism of building with Dockerfile very nicely.

Also, you didn't answer my question.  Why do you consider the simple installation of a 3rd party debian package (or packstack) to be "runtime configuration"

**Comment by cpuguy83** [comment]
@ctindel Portability. Just because something can be done doesn't mean it is supported, and including it in `build`, making it convenient for everyone to do, it means it is supported.
We _will_ be flooded with issues of images not working between hosts... which defeats basically the whole reason for Docker. 

If a package requires `--privileged` to even be installed, then it should be addressed with the package. Installation should not require `--privileged`... and if it really does require it, then this is a sign that the installation itself is not portable (requires changing things on the host)... I'd like to even see docker be runnable in a container without `--privileged` (note that's runnable, you can install it all you want with no issue without --privileged).

**Comment by ctindel** [comment]
But allowing it to be done via docker commit means it is also supported!

I don't understand, you're making tons of people work around limitations in this product because you're worried someone is going to complain to you guys personally about some sort of unsupported image?

You still didn't answer my question about why the act of installing a package (I'm not even talking about configuring it here) is an act of "runtime configuration".  Just saying "portability" doesn't mean anything.

Is there something x86_64 specific about docker?  Won't there eventually be docker images that are built for a specific CPU architecture?  Doesn't that also make them non-portable?  I mean this whole idea that somehow you'll always be able to take any docker image and run it on any docker host in the world regardless of tons of other variables seems impossible anyway, so I don't understand the super strong need to push back on this particular feature that people are asking for.

BTW, let me thank you here for your replies and continued engagement. Plenty of other github projects where issue threads are ignored!

**Comment by PonderingGrower** [comment]
I concur with the point about people working around this by using `docker run --privileged` with `docker commit`. I've made a solution like that for two companies so far. People WILL make images like that and there is no point in acting as if doing it is some horrible, horrible thing.

Hell, if you are so afraid of supporting containers that were built with `--privileged` then just state so clearly in the documentation so people are perfectly aware they are doing it at their own risk. Though I haven't seen any negative effects so far. But then again we haven't been attempting to run containers on different distros.

**Comment by justincormack** [comment]
@PerilousApricot what is actually causing the problem with packstack? We are happy to fix specific problems, or help upstream fix them, but do not think that just adding `--privileged` which gives complete unfettered root access to your host server is the right way to do this. All cases that I am aware of where people have raised specific build issues generally can be fixed, as most things do not actually need root access on the host machine in fact in order to do a build.

**Comment by ctindel** [comment]
@justincormack What's the solution for a 3rd party package (i.e. can't be changed) that starts it's own service where the init script needs to mount a tmpfs filesystem?  I mean even ignoring --privileged for now there's also no way to do a --cap-add or a --security-opt apparmor:unconfined during a build (I don't think?)

**Comment by cpuguy83** [comment]
@ctindel It shouldn't be trying to mount a tmpfs on installation. If it needs tmpfs at runtime then great, but at install time it is most certainly not correct.

**Comment by ctindel** [comment]
@cpuguy83 You are imposing architectural and implementation philosophy on something that is unchangeable because it's coming from a commercial 3rd party.  Not everything is going to be fixed "upstream", even if they fix the newer version of the .deb or .rpm it doesn't mean they'll go back and repost older version which is what you might need in the docker image.

That's the point of this whole discussion, you are imposing arbitrary restrictions that make it much more difficult to use docker because of some philosophical concern about requests for support from people who "do it wrong".

That's like saying Operating Systems should not allow people to change process scheduling classes because if you do it wrong it can lead to priority inversion.  Or, that nobody should make a hammer because it's possible to hit your thumb if you use it wrong.

As has been said many times, docker ALREADY SUPPORTS THIS via the commit command, it's just more painful for your users.  The people who don't want this feature won't use it, and the consenting adults who want to use it by understanding the limitations can do so with eyes wide open.

**Comment by cpuguy83** [comment]
@ctindel More like no, you can't handle this nuclear bomb because you could kill everyone in a 50 km radius.

What is it about this package that it needs to load a tmpfs during installation? Installation is literally extracting files from some archive format to the rootfs.

Anything can be changed.
It is a far simpler and safer change to be made upstream to not mount a tmpfs at installation than it is to enable privileged on build.

Docker is about workload portability. Enabling privileged on build (or extra privileges, or tweaking security profiles, etc) fundamentally breaks this and is not something we can accept today.

`commit` and `build` are two very different things, and just because it's possible to do something one way doesn't mean we should allow doing it in every other way too.

**Comment by cpuguy83** [comment]
```Dockerfile
FROM python

ENV PACKSTACK_VERSION 7.0.1
RUN cd /opt && git clone https://github.com/openstack/packstack.git \
  && cd packstack \
  && git checkout $PACKSTACK_VERSION \
  && rm -rf .git \
  && python setup.py install
```

No privileges required.

**Comment by zerthimon** [comment]
The church of protability.
One day "forced" portability will kill this project - it's already doing it.
So many features are being denied because of elusive portability, so much progress is not being made because of it .....
One day someone will fork a project and make portability optional. Dreams ... dreams .... Amen.

**Comment by tlbtlbtlb** [comment]
If we break it down to two cases:
  1. Installers that use privileged operations frivolously, like mounting tmpfs for performance. Such installers could easily be fixed (but might not be in the near future).

  In this case, it's a valid philosophy for Docker to push back on poorly behaved installers. Most installers have some kind of workaround which just makes the Dockerfile a bit longer.

  2. Installers that depend fundamentally on privileged operations, like installing kernel modules for GPU drivers. These are also fundamentally non-portable. They won't work on docker-machine for Mac, for example.

  In this case, the Docker experience is broken anyway. I can't use docker-machine on Mac, for example, I can only build the image on a compatible target host machine. My use case was installing the nVidia GPU drivers on a host OS (CoreOS) which discouraged installing directly in the host OS.

So, I guess I've come around to see the virtue of not supporting --privileged in either case. I think what made me change my mind is the convenience of building images on my laptop using docker-machine, rather than first pushing my code to a Ubuntu box and building there.

**Comment by ctindel** [comment]
@tlbtlbtlb I don't understand what "virtue" you're referring.  Consider something that is not frivolous, but there are tons of docker images that will run in one environment but not another.  For example, you can mount a host volume into a mongodb container going from linux->linux because and the mmapv1 storage driver will work fine, but you can't pass a mac osx directory through virtualbox to a mongodb container on your laptop because the mmap stuff work work properly in that case.

I realize this isn't a problem with respect to building, but the idea that docker images are "portable" and can "run anywhere" is total nonsense at this point.  If it's the case that they can't run anywhere, what is the virtue in saying they should be able to "build anywhere"?

**Comment by cpuguy83** [comment]
The point is that the mongodb image works everywhere. Providing invalid runtime configuration is a different beast.

Docker has a very specific and intestinal separation of portable configuration and non-portable configuration.

**Comment by ctindel** [comment]
It's such a strange thing to be so concerned with architectural purity regarding Dockerfiles while the product already allows people to work around the limitation by doing a docker commit.  It's like feeling better by not thinking about the fact that your wife had sex with other people before you were married, instead of just realizing there's nothing wrong with her having done it in the first place.

**Comment by boskiv** [comment]
How about this ?
I need to have my real ips inside container to my nginx configuration check pass.

this is my Dockerfile:
```
FROM ubuntu:14.04.4

RUN apt-get update
RUN apt-get install -y software-properties-common
RUN add-apt-repository ppa:nginx/stable
RUN apt-get update
RUN apt-get install -y nginx-full vim
RUN ifconfig lo:0 192.168.168.70 netmask 255.255.255.0 up
RUN ifconfig lo:1 192.168.168.57 netmask 255.255.255.0 up
RUN ifconfig lo:2 192.168.168.58 netmask 255.255.255.0 up

ADD . /etc/nginx

```

```
➜  nginx git:(ha-node-01) ✗ docker build -t nginx4test .
Sending build context to Docker daemon 976.4 kB
Step 1 : FROM ubuntu:14.04.4
 ---> 90d5884b1ee0
Step 2 : RUN apt-get update
 ---> Using cache
 ---> eea42cb6135d
Step 3 : RUN apt-get install -y software-properties-common
 ---> Using cache
 ---> 9db86ab17850
Step 4 : RUN add-apt-repository ppa:nginx/stable
 ---> Using cache
 ---> 5ed2266a93a9
Step 5 : RUN apt-get update
 ---> Using cache
 ---> 09fcfdc1fed3
Step 6 : RUN apt-get install -y nginx-full vim
 ---> Using cache
 ---> cc0c1662e009
Step 7 : RUN ifconfig lo:0 192.168.168.70 netmask 255.255.255.0 up
 ---> Running in 5d962ec4e35d
SIOCSIFADDR: Operation not permitted
SIOCSIFFLAGS: Operation not permitted
SIOCSIFNETMASK: Operation not permitted
SIOCSIFFLAGS: Operation not permitted
The command '/bin/sh -c ifconfig lo:0 192.168.168.70 netmask 255.255.255.0 up' returned a non-zero code: 255
```

Of cause if I run container with privileges option I can settle up ip to loopback interface. But it is one more script to add.

**Comment by MatthewHerbst** [comment]
@cpuguy83 I have about 20 lines or so of iptable entries I'd like to RUN in my Dockderfile but can't since I need `--cap-add= NET_ADMIN`. These commands should happen regardless of who is running the container and regardless of what machine they run it on (container run an internal app). Where/how would you suggest I do that based on what you discuss above?

**Comment by cpuguy83** [comment]
@MatthewHerbst Unfortunately iptables rules won't/can't persist with the image.

**Comment by MatthewHerbst** [comment]
@cpuguy83 I'm on CentOS and can run run `/sbin/service iptables save` to persist the rules to the filesystem. I believe their is a similar capability on Ubuntu and others via the [iptables-persistent](https://packages.debian.org/wheezy/iptables-persistent) package.

**Comment by justincormack** [comment]
You can just generate the iptables rules for that file, there is no need to
actually apply them. The network situation where the container is run may
be very different so you should just apply rules at run time (if that, you
may be better off with the host generating them).

On 16 May 2016 16:03, "Matthew Herbst" <notifications@github.com> wrote:
>
> @cpuguy83 I'm on CentOS and can run run /sbin/service iptables save to
persist the rules to the filesystem. I believe their is a similar
capability on Ubuntu and others via the iptables-persistent package.

**Comment by MatthewHerbst** [comment]
@justincormack don't know why I didn't think of that! Thanks!

**Comment by nostrebor** [comment]
How are you supposed to execute commands that require privilege when using `docker service`? I need to set the hostname on a few of my machines, but unfortunately this requires privilege.

**Comment by cpuguy83** [comment]
@nostrebor that is very much unrelated to this open issue.
We are evaluating what options need to exist for services rather than copying them 1-to-1. Privileged mode probably will not be in 1.12 for services.

**Comment by drstapletron** [comment]
I'm trying docker builds which compile something for installation, but it must compile against libraries which exist on a CVMFS network file system.  Of course, I can't mount CVMFS without running --privileged, so I can't do it at all using a Dockerfile.

@cpuguy83 @tlbtlbtlb This is a case of an 'installer' depending fundamentally on privileged action.  But it's not what I would call a 'frivolous use', I just need access to shared libraries on a network file system.  Still, installation in this case is NOT simply extracting files from some archive.

I don't understand how mounting a network file system is a portability problem.  (All of our target environments will have access to this file system.  They are required to because they build other code which ALSO must link to binary code on the network file system.)

Should I try a different approach?  Should I mount CVMFS on the host and share that directory with the container or something?  I don't want to have to set up an external build system just to make this image - the image upon which it's based already has an entire build system that will do the job.  I just need to be about to mount CVMFS.

I was excited about doing this with a Dockerfile, but it seems I'll have to do it the hard way using some scripts with `docker run --privileged`, or use something else instead of docker.  Is there a way of mounting a filesystem in a container *without* having privileged access?

**Comment by iahmad-khan** [comment]
I did a workaround , by placing / echoing privileged commands inside a script and used CMD instruction to run the script at container entry point , so after building an image like that , i am able to just run the container in privileged mode and everything works.

**Comment by iahmad-khan** [comment]
@drstapletron  , according to CERN cvmfs documentation , you have two options for now , either mount cvmfs from host to container or install cvmfs inside a privileged container. 

For the seconds case , i have just write a docker-file for cmssw guys , here:
https://github.com/iahmad-khan/system-admin/blob/master/cvmfs-inside-docker.Dockerfile

so using this file , you can just build an image ( or may be you grab it from cmssw dockerhub) , and run it in P mode , and everything will already be there inside container ( ls /cvmfs/*)

**Comment by andyrj** [comment]
Not sure if this was covered above as it is a rather long list of feedback on this issue.  I too would like to have --privileged build commands.  My current use case is to work around an issue I ran into with building the go ebuild on a gentoo stage3.  If I don't use a container, the current instructions in the gentoo handbook cause systemd to behave erratically once I' umount -l /mnt/gentoo/proc,sys,dev' from a system booted with systemd...  When I move my stage3 build into a docker container everything works fine up until a build requires ptrace, or some other restricted feature...  which the go-1.7.1.ebuild seems to require.

For now I am simply running the build within a docker run command, committing and then continuing from there, but it would be preferable to enable ptrace within the docker build command itself to avoid the manual steps.

**Comment by seltzy** [comment]
I'd like this feature as well. I want to create a build environment, but it requires a kernel module and modprobe won't cooperate with me when I'm building. Are there workarounds for this?

Specifically:
```
modprobe vcan && \
ip link add type vcan && \
ifconfig vcan0 up
```
Seems like a totally reasonable use-case.

**Comment by tamsky** [comment]
@seltzy I suggest not holding your breath waiting for anyone from `docker` to acknowledge the reasonableness of your use case.

When it comes to architecture and feature-inclusion, they are very heavy-handed, pragmatic, aloof, and tend to ignore any and all use cases that don't fit into _their_ roadmap.

Us ordinary folks need to understand, the `docker` team makes architecture decisions that further their own (likely business customer and self-serving) needs, and those decisions rarely overlap with our (the public end-user) very laboriously crafted [issues](https://github.com/docker/docker/issues) that we file here.

They are, of course, free to allocate their engineering resources in this way.
But it does provide a track record that would beg the imagination if the company was ever described as "taking user's needs seriously."

**Comment by zerthimon** [comment]
@tamsky you nailed it !

**Comment by cpuguy83** [comment]
@tamsky I can understand why you would think that, since the project has not accepted a feature that you clearly desire.

I can assure you this has nothing to do with any sort of business decision. The fact is, `--privileged` on build will yield to non-portable images.
Things like `modprobe` in the build environment are not helpful and can even cause two builds to yield entirely different results.

I myself have implemented `--privileged` on build. It is not an engineering problem, and really it's quite trivial to implement. It's supporting it that is the problem.
For advanced users, a custom builder can be implemented, even reusing the existing code, using the existing API's that can include privileged support.

That said, this issue is still open for a reason. It is because people are listening.

Thanks for your consideration.

**Comment by seltzy** [comment]
@cpuguy83 Thank you for the explanation. I didn't realize it was a portability problem. I suppose my desire for this is fueled by misunderstanding.

What's the general philosophical approach to take when faced with the temptation to have privileged build?

**Comment by drstapletron** [comment]
@seltzy don't be so sure that your use case isn't a reasonable example of the need for this feature

@cpuguy83 I'm still [waiting for a reply](https://github.com/docker/docker/issues/1916#issuecomment-239967893) about my use case.  Our institution's build system is distributed over a network file system which I **must** mount in my container.  This requires the container to run in privileged mode.  My institution's decision to use a network file system for software distribution is not unusual for particle physics.  You claim that `--privileged` on build creates non-portable images, but this is completely unrelated to my use case.  Our development model has already given up any portability we *might* lose due to our use of a network file system (really?).  We just need the development machine to be about to mount it.

@cpuguy83 PS you mentioned a custom builder.  Where can I find information about this?  Thanks!

**Comment by ctindel** [comment]
This whole discussion about container portability is a giant red herring anyway.  You can already accomplish the same thing by creating a stage 1 image, launching a container in privileged mode, doing whatever you have to do, then using docker commit to create the final image from that container.  

Once they added the docker commit option any notion of image portability went out the door anyway, so forcing people to do this in 3 steps instead of 1 is not gaining anything and only serves to annoy people who could really use a privileged build option.

**Comment by cpuguy83** [comment]
@drstapletron Mounting a filesystem is not necessarily something that can break portability (unless someone was expecting that to be mounted in the image).
The issue here is having the capability to mount a filesystem also means being able to do a bunch of other nasty things.


@ctindel Yes, you can do anything you want to in the containers you create. The fact that `docker build` is "the" supported method of building images, though, means we need to ensure that images built with it *are* portable.

Portable images are not a red herring. Workload portability is the primary design goal of Docker. Our prime directive, as it were. 

@seltzy Most things which require extra privileges belong at runtime, because most of the time the elevated privileges are used to modify the host in some way.
That said I can certainly understand that some things are required at build time (such as the nfs mount above).... but even with the case of NFS and manually building the image (not with `docker build`), I would not give the container `--privileged` or any extra capabilities at all, instead I would mount the nfs export as a volume.

**Comment by justincormack** [comment]
@drstapletron `mount` does not require `--privileged` only a more restricted capability set and is much more likely to happen sooner than a full privileged mode, as that gives full root access to the host which most people do not want. (It still has security issues but they are more manageable).

**Comment by lamawithonel** [comment]
So I have a completely portable, non-host-modifying use case.  It is even open source and you can see it [here](https://github.com/lamawithonel/simp-builder-docker).

Basically, I want to run Mock in a portable Docker container to build a customized CentOS ISO image.  Mock, for those that don't know, is a containerized RPM builder.  The problem is, since it uses containers I need `--privileged` or `--cap-add`.  Ideally, I think `docker build` would work like a function, taking a few arguments and returning its final result.  Without these flags, though, I cannot do that.

**Comment by gboddin** [comment]
Same here ! Using mock inside docker is nightmare because of that :(

```
Sending build context to Docker daemon 9.728 kB
Step 1 : FROM centos
 ---> 980e0e4c79ec
Step 2 : MAINTAINER Gregory Boddin
 ---> Using cache
 ---> 93e709c87f25
Step 3 : RUN yum install -y spectool mock
 ---> Using cache
 ---> 7006ef8d0276
Step 4 : RUN useradd mock -g mock
 ---> Using cache
 ---> bfb931c56d89
Step 5 : ADD *.cfg /etc/mock/
 ---> Using cache
 ---> 15521d2822b1
Step 6 : RUN su mock -c"mock -r edge-5-x86_64 --init"
 ---> Running in 02ed26deec87
ERROR: [Errno 1] Operation not permitted
ERROR: The most common cause for this error is trying to run /usr/sbin/mock as an unprivileged user.
ERROR: Check your path to make sure that /usr/bin/ is listed before /usr/sbin, or manually run /usr/bin/mock to see if that fixes this problem.



 ---> 1e9c7254e4dc
Removing intermediate container 02ed26deec87
Step 7 : RUN su mock -c"mock -r edge-6-x86_64 --init"
 ---> Running in e62ef5d67e38
ERROR: [Errno 1] Operation not permitted
ERROR: The most common cause for this error is trying to run /usr/sbin/mock as an unprivileged user.
ERROR: Check your path to make sure that /usr/bin/ is listed before /usr/sbin, or manually run /usr/bin/mock to see if that fixes this problem.



 ---> 32867e0755c2
Removing intermediate container e62ef5d67e38
Step 8 : RUN su mock -c"mock -r edge-7-x86_64 --init"
 ---> Running in 00497e9ab06d
ERROR: [Errno 1] Operation not permitted
ERROR: The most common cause for this error is trying to run /usr/sbin/mock as an unprivileged user.
ERROR: Check your path to make sure that /usr/bin/ is listed before /usr/sbin, or manually run /usr/bin/mock to see if that fixes this problem.



 ---> 36b2942a07c6
Removing intermediate container 00497e9ab06d
```

**Comment by jtmoon79** [comment]
@cpuguy83 wrote:

> The fact is, --privileged on build will yield to non-portable images.

But why not allow `--privileged` for those not requiring far-reaching portability?
A simple note in the official documentation would be a reasonable compromise (e.g. _Warning: passing`--privilege` to the `build` command may result in a less portable image!_)

**Comment by stuaxo** [comment]
It could be called  ```--non-portable```.   I haven't used the deployment parts of docker yet, but the isolation + overlay filesystem stuff has been really useful without that.

**Comment by robhaswell** [comment]
I need to use some proprietary software which requires a privileged container to install. There's nothing I can do about this. I'm stuck with needing to do the 3-stage build, run, commit process.

Container portability doesn't mean anything to me or my business, in fact I'll bet it doesn't mean anything to the vast majority of businesses. What is important is that I would like to maintain less software, so I think selecting portability over usability in this issue is detrimental to Docker.

**Comment by skshandilya** [comment]
+1 for this, in the build process we use setfacl, this fails during build and services fail to start in the container

**Comment by samsun387** [comment]
+1 for this. In our build process it's required to mount the /proc and /dev. Ideally we should the mount step as part of the dockerfile.

**Comment by cpuguy83** [comment]
@samsun387 Why does your build process require this?

@skshandilya setfacl is not portable and I would be surprised if acl's could be persisted to an image.

@robhaswell "requires a privileged container" doesn't help much. What is actually being used on install?

**Comment by Betriebsrat** [comment]
+1 for this. mock init needs this.
almost read the whole issue. don't get why people are asking "why do you need this" for 3 years straight.

**Comment by cpuguy83** [comment]
@Betriebsrat Because "X needs this" is not really all that helpful.
What is "X" doing? Why does "X" need this during the build phase?

For instance, the above case with capabilities to mount `/proc` and `/dev` really do not seem like the right place for the build phase, and even seem like the image would be tied to the host in such a case.

"privileged" is also a tank. It opens up absolutely everything, disables all security features, gives write access to typically read-only places... where someone probably just needs to be able to do a very specific thing.

These questions are asked so we can get the real use case and how we might satisfy such a case.

By the way, when I say "security features" I mean two things:
1. Things to prevent hacking
2. Isolation of application concerns from host concerns (ie, image build should not tie the image to the host it's built on).

**Comment by stuaxo** [comment]
Looks like mine was solved by 21051, I'm out, for now :)

**Comment by tamsky** [comment]
@shykes said on Nov 28, 2013 @ https://github.com/docker/docker/pull/2839#issuecomment-29481246 ::
> Sorry, the current design is to enforce '1 source, 1 build' which is why we don't allow any arguments to docker build other than the source directory.
> 
> I understand the fact that some builds currently can't take place because they need privileged operations. To handle that properly we might need to 1) allow a Dockerfile to express the need to be built in a privileged way, and 2) implement an authorization/trus system which allows docker to pause the build, properly warn the user of the risk, expose information on the origin and trustworthiness of the Dockerfile, and then collect the user's decision to allow or deny the build.

@cpuguy83 , is has the design changed at all from enforcing: "1 source, 1 build" ?
Is the Docker Project willing to alter that design and allow this community-requested feature?

Shykes comment above appears to spell out what "we might need to" do to get this handled.  At the same time, the language used ("might") appears to provide the docker project lots of room to come up with additional reasons to reject this design change.

**Comment by ctindel** [comment]
Adding a a NEEDS_PRIVILEGED  declaration makes sense but all this stuff about pausing the build? Just fail with an error and let the operator pass the --privileged option if they do indeed want to allow the privileged build.

**Comment by PonderingGrower** [comment]
@cpuguy83 the thing is, people who ever need privileged mode in build are usually power users that know perfectly well what the risks of using it are. And most of them accept that and work around this by simply using `docker commit` as part of their build for the step that needs it.

You are not in any way preventing people from using privileged mode in build, you are just making it annoying to do.

If your goal is making it annoying to do then just say so outright and close this issue instead of letting it go on for years and years.

Say "we won't fix this because we want it to be annoying to do" and close this issue.

/thread

**Comment by lamawithonel** [comment]
@cpuguy83, from my understanding, Mock uses `unshare(2)` with the `CLONE_NEWNS` flag-- and possibly others-- when it creates its chroot/container environment. This requires at least `CAP_SYS_ADMIN`.

**Comment by robhaswell** [comment]
> What is "X" doing? Why does "X" need this during the build phase?

In our use-case, we don't know. It's some propietary piece of crap that we can't change. The thing is, our business doesn't care about "security" (in this context) or portability, or any of the concerns that have been listed. We just want to put this damn piece of crap in a container and move on to doing something valuable.

As @PonderingGrower says, we're going to do it anyway, it's just a matter of how much time we waste while doing it.

**Comment by thaJeztah** [comment]
> people who ever need privileged mode in build are usually power users that know perfectly well what the risks of using it are

I strongly disagree with that assumption. Overall, people that use `--privileged` are the same category of users that blindly run `chmod -r 777` "because someone wrote that it fixed the issue"

> In our use-case, we don't know. It's some propietary piece of crap that we can't change. The thing is, our business doesn't care about "security" (in this context) or portability, or any of the concerns that have been listed.

"in this context" here, meaning: giving "some proprietary piece of crap" root access on your host.

**Comment by robhaswell** [comment]
@thaJeztah I have no problem with that. It's software we have purchased and pay support for. If we weren't using containers, it would still need root to install.

**Comment by FromZeus** [comment]
We need this feature for using dind while building to preconfigure some containers inside the one we are building.

**Comment by andrew-aladev** [comment]
`--cap-add=SYS_PTRACE` will fix glibc compilation in docker.

**Comment by andrew-aladev** [comment]
What are you discussing here for 3 years?

`docker run` has an options `--cap-add`, `--cap-drop` and others. So `RUN` command in `Dockerfile` want to have the same options. So `Dockerfile` want to send requests to the parent machine and ask it to add/drop some privileges.

Parent machine can do what it want with this request. You can make shell interactive, you can make a gui confirm dialog, etc. Why are you discussing the resolution of these requests in this issue?

**Comment by ctindel** [comment]
A significant number of docker users want the ability to --cap-add or --privileged in the build command, to mimic what is there in the run command.

That's why this ticket has been open for 3 years with people constantly chiming in even though the maintainers aren't interested in giving the users what they want in this specific instance.

**Comment by andrew-aladev** [comment]
@ctindel This is definitely the problem of this issue. There is a gap between `docker build --cap-add` and `RUN --cap-add`.

Some people want to resolve privileges requests from child machine with just `docker build --cap-add=caps_array`. What is it? This is just: `caps_array.include? requested_cap`.

Some people want `pre_requested_caps.include? requested_cap`. Some people want `stdout << requested_cap, stdin.gets == 'y'`.Some people want `gui_confirm requested_cap `. Some people will definitely want `UAC_fullscreen_dialog requested_cap`.

The resolution method of `requested_cap` depends on user taste and I think this question won't ever be done.

But `RUN --cap-add` has nothing to do with people tastes. What are we waiting for?

**Comment by ctindel** [comment]
@andrew-aladev I don't really understand what your post is saying.  The point is that people have 3rd party software (RPMs, DEBs, whatever) that is not under their control, which they want to install into an image at "docker build" time, and which requires extra capabilities in order to install correctly.  Since they are third party RPMs there is no way to resolve the requirement for increased privileges during the install phase.

They are working around the problem by running a container with those increased capabilities, installing their software, and then creating an image from that container.  It's a pain and clearly shows that there is no functional reason to prohibit cap-add at build time since the same end can be achieved circuitously.

**Comment by andrew-aladev** [comment]
@ctindel My english is not very good, sorry.

I know. I've tried to emerge glibc and I've received `ptrace not permitted`.

`docker` can run container with increased/decreased capabilities by itself. `RUN` command in `Dockerfile` should support `--cap-add`, `--cap-drop`, etc.

Lets imagine that our `Dockerfile` will have `RUN --cap-add=SYS_PTRACE -- emerge -v1 glibc`. How it would work?

1. Child machine sends a request to parent and asks for `SYS_PTRACE`.
2. Parent allows extended capabilities.
3. Parent creates new container with SYS_PTRACE allowed.

I see that nobody in this issue is actually arguing with that. People are just arguing about a **method** of allowing these capabilities.

@thaJeztah said
> Overall, people that use --privileged are the same category of users that blindly run chmod -r 777

This man wants a more flexible **method** of validation required capabilities that just `log :info, requested_cap; return privileged?`.

@ctindel you said
> Adding a a NEEDS_PRIVILEGED declaration makes sense but all this stuff about pausing the build? Just fail with an error and let the operator pass the --privileged option if they do indeed want to allow the privileged build.

You want to make shell interactive. You want `stdout << requested_cap, stdin.gets == 'y'`. This is the another **method** of validation required capabilities.

@cpuguy83 said
> someone probably just needs to be able to do a very specific thing... things to prevent hacking.

This man wants `docker build --cap-add=caps_array` `caps_array.include? requested_cap`. This is the another **method** of validation required capabilities.

So I am asking: Why `RUN` in `Dockerfile` still have no support of `--cap-add`, `--cap-drop`, etc? Nobody argues about that. 3 years passed!

**Comment by mlalpho** [comment]
@andrew-aladev I assuming no one has argued for that syntax because it's been made clear that dockerfile syntax is frozen until the builder is rewritten/refactored/decoupled from the main engine. https://github.com/docker/docker/issues/29719#issuecomment-269342554

More specifically, the title of the issue and OP are requesting --privileged build

**Comment by mbana** [comment]
this gets a Fonzie thumps up: https://avatars1.githubusercontent.com/u/7646183?v=3&s=88.

being able to run `strace` in the `build` step helps a lot.
currently, i work around this by moving all the things i need to debug to the `run` step - not ideal.
does anyone know why it would work in the `run` and not the `build` step? i.e., the the historical reasons.

**Comment by sneumann** [comment]
There is a proposed solution/workaround to this in 
https://github.com/docker/docker/issues/6800#issuecomment-50494871 :

> if you have issues in docker build, you can use a "builder container": 
> `docker run --cap-add [...] mybuilder | docker build -t myimage -`

Could someone (possibly @tiborvass) elaborate on this ? What kind is `mybuilder` here ?
The image name with some ENTRYPOINT ? Or is the image part of `[...]` and `mybuilder` refers 
to a shell script ? And how do I convince `docker run` to pass the context.tar.gz into the `docker build -` 
if that is really what's happening here. Thanks in advance, Steffen

**Comment by tiborvass** [comment]
@sneumann `mybuilder` would be an image name and have some `CMD` or `ENTRYPOINT` indeed. The contract for that workaround to work is that `mybuilder` will have to `tar` the context from within the container and let it go to stdout. That is passed on to `docker build`'s stdin, thanks to the shell pipe `|` and is considered to be the context because the context path for `docker build -t myimage -` is `-`.

**Comment by mbana** [comment]
a bit strange upon looking at the code it seems that this option is avaialble in the `build` command:

* `run`: i assume that `container` here means `run` - https://github.com/docker/docker/blob/bb0a532fc21cb6b9390fb7e5eb0054bdc8045bbc/cli/command/container/opts.go#L485.
* `build`: https://github.com/docker/docker/blob/bb0a532fc21cb6b9390fb7e5eb0054bdc8045bbc/cli/command/image/build.go#L108.

anyone more clued up have any idea why it's not being applied?

**Comment by thaJeztah** [comment]
@mbana the `--security-opt` is for native Windows containers, which support "credentialspec" https://github.com/docker/docker/pull/23389

**Comment by mbana** [comment]
is it possible to modify this and make it persist so that future `build` will enable `ptrace`?

for anyone interested here are some good links:

* http://www.projectatomic.io/docs/docker-building-images/ - how to build the image using the technique described above. i.e., use `run` to `build` the image.
* https://blog.afoolishmanifesto.com/posts/how-to-enable-ptrace-in-docker-1.10/ - misc

**Comment by scjody** [comment]
I've seen a lot of claims from various folks that this feature isn't needed because builds can be changed to not require certain privileged operations, but no suggestions on what to do about the "docker in docker" case. If a build needs to run `docker` commands, for example to pull down some  images that we want to ship inside this one, or to build a sub-image, how are we supposed to do this without some sort of privileged build option?

For now I'm going to work around this using `docker run` and `docker commit` but it would be great if `docker build` could support this use case.

**Comment by cpuguy83** [comment]
@scjody Sounds like you want #31257

**Comment by scjody** [comment]
@cpuguy83 I'm not sure that covers what's going on in this case but I'll give it a shot once that's merged. Thanks!

**Comment by onlyanegg** [comment]
Hi, i'd like to throw my name into the please-implement-this hat. Or maybe there's a different solution to my problem (docker noob here) that someone could point me to? 

I'm trying to build an image based on the official [centos/systemd](https://hub.docker.com/r/centos/systemd/) image and provision it with Saltstack. This requires starting (and maybe restarting) the salt-minion daemon with systemd which can't be done (AFAIK) without privileged mode.

**Comment by thaJeztah** [comment]
@onlyanegg I think in that situation, Saltstack largely replaces the functionality of the builder; keep in mind that each `RUN` statement is executed in a new container, at which point the previous build container is stopped, and committed to an image / layer.

Have you considered performing the build by running a container, and committing the results (`docker commit`)?

**Comment by onlyanegg** [comment]
Thanks for responding, @thaJeztah. I didn't realize that's what the `RUN` directive did. I did read most of this issue, so I am aware of the `docker build` -> `docker run` -> `docker commit` workaround, which is what I'll likely end up doing. I'm just more in favor of having a single file describe my image - seems neater. Maybe I can put all those steps in packer post-processors and then I'll have that.

**Comment by derberg** [comment]
Why is this one ignored so much? In times of containers, kubernetes and minikube, and usage of docker in CI and development environment unification, this functionality is really crucial.

**Comment by jpetazzo** [comment]
@onlyanegg you should be able to restart services _without_ privileged mode. If you have a Dockerfile illustrating that (i.e., "the `RUN` command at line 8 of this Dockerfile doesn't work because it requires privileged mode") I would be more than happy to take a look!

@derberg precisely! In times of containers, CI, CD, it is important that build tools can be contained (in the security sense). If you allow privileged mode, you have to change dramatically how you use CI tools like Jenkins, Travis, Codeship, etc. Same question: if you have a Dockerfile that requires privileged mode, I would be happy to take a look to suggest alternatives.

Thank you!

**Comment by derberg** [comment]
@jpetazzo try to get a docker image with docker inside it:

```
FROM ubuntu:16.04

# Get dependencies for curl of the docker
RUN apt-get update && apt-get install -y \
    curl \
    sudo \
    bash \
    && rm -rf /var/lib/apt/lists/*

RUN curl -sSL https://get.docker.com/ | sh
```

Now build it and start it. After starting run `service docker start` to start the docker deamon. Then check the status of the service `service docker status`:
- with privileged flag status is ok and you can start container without issues
- without the flag, it never starts

**Comment by derberg** [comment]
@jpetazzo ech, just noticed you are the creator of https://github.com/jpetazzo/dind :) so you are aware of docker in docker concept :)

Anyway, so you are aware that privileged flag is needed for run. So now you can imagine a group of people working on some environment and wanting to have a unified environment for development with already some preconfigured things inside, for example minikube with preinstalled components or anything else

**Comment by zerthimon** [comment]
So, if there a way to mount an NFS or SMB share in `docker build` yet ?

**Comment by thaJeztah** [comment]
@derberg those steps won't work, even if the build container was running `--privileged`; the docker packages (and install script) are (for example) installing kernel packages on Ubuntu 16.04. 
That's exactly the reason `--privileged` is a bad idea for `docker build`, because it would be able to make changes on the _host_.

Even though docker will need privileged when _running_, the installation itself does not need this; for example, here's the steps you'd run to install docker in your image;

```bash
docker build -t foo -<<'EOF'
FROM ubuntu:16.04

RUN apt-get update && apt-get install -y \
    apt-transport-https \
    ca-certificates \
    curl \
    software-properties-common \
    && rm -rf /var/lib/apt/lists/*

RUN curl -fsSL https://download.docker.com/linux/ubuntu/gpg | apt-key add -
RUN add-apt-repository "deb [arch=amd64] https://download.docker.com/linux/ubuntu xenial stable"
RUN apt-get update && apt-get install -y docker-ce \
    && rm -rf /var/lib/apt/lists/*
EOF
```

And you can run it just fine (I'm using `--privileged` here, but perhaps more fine-grained permissions are possible):

```
docker run -it --rm --privileged -v /var/lib/docker foo dockerd --debug
```

**Comment by kraml** [comment]
Here is how I circumvent this issue, for people who really need to build a docker image in privileged mode. It may not solve all cases but can help in some.

This method need a Dockerfile, a docker-compse file, and a shell script.

## The Dockerfile
It is the same as you would need to build the image. The only difference, stop at where you need do privileged operations, don't include them. They need to be run by the docker-compose, as a script. For example:

```Dockerfile
FROM ubuntu:16.04
RUN apt-get update && apt-get install <your packages>
# And more commands
......

## Below are the operations you intended to run in privileged mode when building the image, which does not work.
# More commands....
## But they now are moved to a separated shell script and it will be included in the image
COPY further-commands-to-run-in-privileged-mode.sh /
```

Those more commands that need to be run in privileged mode are now in ```further-commands-to-run-in-privileged-mode.sh```. It is included in the image and later will be run by docker composer to finished the build process.

## The docker compose file
The compose file is the key. It will build the image first from above Dockerfile, and start a container from that image **in privileged mode**, then you can do your privileged operation there. For example:

```yaml
version: '3'

services:
  your_service:
    container_name: your_container
    # First build the image from the Dockerfile
    build:
      # Change this to where you keep above Dockerfile
      context: ../docker-build
    image: "your_image_name:your_image_tag"

    # Then start a container from the just built image in privileged mode to finish what's left
    entrypoint: /further-commands-to-run-in-privileged-mode.sh
    privileged: true
```

## Build the image
Below steps can also be save in a shell script.

```bash
# First build the image and container(in privileged mode)
docker-compose -f docker-compose.yml up

# Then commit the temporary build container to a new image, change the ENTRYPOINT to what you want
docker commit \
    -c 'ENTRYPOINT ["/bin/bash"]' \
    <build container name> \
    <final image name>:<final image tag>

# Remove the temporary build container
docker rm <build container name>
```

**Comment by derberg** [comment]
@thaJeztah I don't have a problem with installation, I have problem with starting the docker service during the build and pulling some images to have them available in the image out of the box.

Add the following script to your Dockerfile and you will see that docker service never gets up 

```
#!/bin/bash

service docker start

sleep 20

service docker status

docker pull busybox
```

**Comment by jpetazzo** [comment]
@derberg OK, I see! _Personally_, if I wanted to include images in a `dind` container, I would download them (for instance with `reg`) and I would load them the first time that the container starts. Why? Because if you pull the images during the build, the image will work _only if `dind` is started with the same storage driver_. In other words, your image might or might not work on other machine.

Also–if your images are big (i.e. anything else than, say, `busybox` or `alpine`) you will end up with a really big DinD image...

I'd love to know more about your final use-case – because I'm sure we can help you to find a more efficient way than baking a huge DinD image :-)

(Otherwise, the solution proposed by @kraml is rather elegant!)

**Comment by cpuguy83** [comment]
Also see https://github.com/moby/moby/blob/master/contrib/download-frozen-image-v2.sh
Downloads images using only bash+curl+tar.
We use it here: https://github.com/moby/moby/blob/master/Dockerfile#L171

**Comment by derberg** [comment]
@jpetazzo we already run with such workaround build-run-commit, but yeah, it is still a workaround from my point of view. Use case is pretty specific related to kubernetes and minikube environment and there is nothing we can do for now. For now we were able to start minikube in docker only with docker as a daemon, using virtualbox or other vm drivers didn't work, so we are dependent on `dind` approach

**Comment by inthecloud247** [comment]
Ran into this issue trying to build a image containing a legacy application (pretty normal use-case), where the installer tried to run sysctl commands and failed. 

Coming back to this thread and reviewing the entire 4 years (!!!) of back and forth on the issue of how to add some kind of privileged capabilities to the docker build command, it appears that the available options in this situation are either a nasty bunch of sed commands to modify the installer to remove the sysctl calls, or a multi-stage build -> run -> commit pipeline. I agree with @derberg that 'build -> run -> commit' feels to be a workaround (imo a gross/hacky workaround) and I don't think that my use-case is that unique. Checking other threads I've seen plenty of people reporting issues with various application and database installations with failed `docker build` commands due to lack of privileges.

At this point, the docker run command supports extensive 'privileged' options, along with "fine grain control over the capabilities using --cap-add and --cap-drop". And so, I think objections on a security or technical basis are moot. If the privileged run option is added along with '--cap-add' and '--cap-drop', a security-conscious engineer could choose to limit a privileged build to only include the specific capabilities required for their build.

**Comment by iahmad-khan** [comment]
Hi , 

I have already reported this before , same problem.

What about those who just want to run one container per VM with same user on VM and container , using docker just as a packaging tool?

Is there still a security concern?

**Comment by krojew** [comment]
Ran into this problem. Could really use capabilities for build.

**Comment by stanzgy** [comment]
Ran into this issue too. 
It's very useful when using docker as CI/CD slaves which could require privileged permission on `docker build` to install the CI/CD build/test toolchains when building the slave docker image.
I'm waiting for this feature for years and really hope it could be supported in future.

**Comment by maneamarius** [comment]
I really don't understand why this there is so much pushback from devs regarding --privileged for docker image.
If the users want to shoot themselves in the foot, why not let them? Just put a warning message and that's it. There are already workarounds for achieving same thing, why not make it easier for the users that really need it??
It's been 4-5 years and has been no progress on this.
Just amazing...

**Comment by maneamarius** [comment]
As of today, not even this feature has been implemented yet:
RUN --cap-add=SYS_PTRACE
which would fix the issue of many users..

**Comment by maneamarius** [comment]
Could you please suggest how I can build this Dockerfile on Gentoo Linux host:

```
FROM gentoo/stage3-amd64
# Download and extract latest portage
RUN wget http://distfiles.gentoo.org/snapshots/portage-latest.tar.bz2 && \
    wget http://distfiles.gentoo.org/snapshots/portage-latest.tar.bz2.md5sum && \
    md5sum -c portage-latest.tar.bz2.md5sum 
RUN tar -xjvf portage-latest.tar.bz2 -C /usr
RUN emerge dev-lang/go
```

because I am getting this error when emerge-ing dev-lang/go:
```
##### Building Go bootstrap tool.
cmd/dist
 * /var/tmp/portage/sys-apps/sandbox-2.12/work/sandbox-2.12/libsandbox/trace.c:_do_ptrace():75: failure (Operation not permitted):
 * ISE:_do_ptrace: ptrace(PTRACE_TRACEME, ..., 0x0000000000000000, 0x0000000000000000): Operation not permitted
/usr/lib64/libsandbox.so(+0xb692)[0x7fd10e265692]
/usr/lib64/libsandbox.so(+0xb778)[0x7fd10e265778]
/usr/lib64/libsandbox.so(+0x6259)[0x7fd10e260259]
/usr/lib64/libsandbox.so(+0x6478)[0x7fd10e260478]
/usr/lib64/libsandbox.so(+0x7611)[0x7fd10e261611]
/usr/lib64/libsandbox.so(execve+0x3f)[0x7fd10e2634ff]
bash[0x41d8ff]
bash[0x41f387]
bash[0x420138]
bash[0x4219ce]
/proc/330/cmdline: bash ./make.bash 

 * ERROR: dev-lang/go-1.9.2::gentoo failed (compile phase):
 *   build failed
 * 
 * Call stack:
 *     ebuild.sh, line 124:  Called src_compile
 *   environment, line 1034:  Called die
 * The specific snippet of code:
 *       ./make.bash || die "build failed"
 * 
 * If you need support, post the output of `emerge --info '=dev-lang/go-1.9.2::gentoo'`,
 * the complete build log and the output of `emerge -pqv '=dev-lang/go-1.9.2::gentoo'`.
 * The complete build log is located at '/var/tmp/portage/dev-lang/go-1.9.2/temp/build.log'.
 * The ebuild environment file is located at '/var/tmp/portage/dev-lang/go-1.9.2/temp/environment'.
 * Working directory: '/var/tmp/portage/dev-lang/go-1.9.2/work/go/src'
 * S: '/var/tmp/portage/dev-lang/go-1.9.2/work/go'
```

**Comment by AMD-NICK** [comment]
How can I run it without `--cap-add=SYS_ADMIN --device /dev/fuse` or `--privileged`?
```Dockerfile
RUN apt-get -y install unionfs-fuse
RUN unionfs-fuse -o cow dir1=RW:dir2=RO dir3/
```

I can do it with a separate bash file in entrypoint, but I need single Dockerfile

**Comment by thaJeztah** [comment]
@amd-nick what's your expectation of the `RUN unionfs-fuse ...` line during build? Even if that worked, it would only have the filesystem mounted during that single `RUN`, and be gone in the next step.

**Comment by AMD-NICK** [comment]
@thaJeztah it's hard to explain for me. I am trying to modify [this repo](https://github.com/suchipi/gmod-server-docker). Can I just skip this line on the building?

**Comment by dn2k-dev** [comment]
Hi

Randomly the docker build choose an hostname starting zero '0' which breaks our application, I tried to run "hostname" in such case inside my DockerFile but faced the same issue.

I also would like to have an option to run the docker build with RUNP or get an option to choose the hostname during build.

**Comment by nelsonjchen** [comment]
Has anybody tried building these kinds of images with [Kaniko](https://github.com/GoogleContainerTools/kaniko)? I just did it with @maneamarius 's Dockerfile on Docker for Mac and it seems to build successfully once you call Kaniko's `docker run` "build" command with `--cap-add=SYS_PTRACE`. Though, I'm having a bit of trouble loading the resulting tarball locally, the RAM usage is a bit high since it can't use overlayfs, and layer caching is still WIP. Things might Just Work if I push to a registry but I haven't tried that yet.

**Comment by avidspartan1** [comment]
Having this feature would greatly help efforts to build Docker images via Puppet on Redhat/CentOS base images.

**Comment by nelsonjchen** [comment]
Since I last posted, I've followed back up with the changes in [Kaniko](https://github.com/GoogleContainerTools/kaniko). They are no longer tarballing in memory and are tarballing onto the disk which means support for Dockerfiles describing big images. Layer caching is still a WIP but they have an option for caching the base images for the moment (That means currently no fast `RUN` iteration save and run kind of work but we can cache `alpine`, `ubuntu`, and whatever popular bases are out there).

It's at a state where I've been successful in building @maneamarius's [Dockerfile that emerges Golang in a Gentoo image](https://github.com/moby/moby/issues/1916#issuecomment-361175319) in this project/demo:

https://github.com/nelsonjchen/kaniko-privileged-maneamarius-moby-1916

I've also configured Azure Pipelines to build the Dockerfile into an image with Kaniko with `--cap-add=SYS_PTRACE`, load Kaniko's output tarball, and run `go version` in the generated image. I figured some interactive "proof of life" would be interesting. A link to the Azure Pipeline build outputs is at the top of the README.

**Comment by alexey-vostrikov** [comment]
Use buildah Luke

**Comment by AkihiroSuda** [comment]
I'm closing this issue, because the feature is now available as `docker buildx build --allow security.insecure`

https://github.com/docker/buildx/blob/master/README.md#--allowentitlement
https://github.com/moby/buildkit/blob/master/frontend/dockerfile/docs/experimental.md#run---securityinsecuresandbox

**Comment by bharathappali** [comment]
@AkihiroSuda I have updated my docker to version 19.03 to try `buildx`. When i was trying the command you mentioned, its giving me and error 
```
$ docker buildx build --allow security.insecure -t sample-petclinic -f Dockerfile .
[+] Building 0.0s (0/0)                                                                                                                                                         
failed to solve: rpc error: code = Unknown desc = entitlement security.insecure is not allowed
```

`Docker version`:
```
Client: Docker Engine - Enterprise
 Version:           19.03.2
 API version:       1.40
 Go version:        go1.12.8
 Git commit:        c92ab06
 Built:             Tue Sep  3 15:57:09 2019
 OS/Arch:           linux/amd64
 Experimental:      true

Server: Docker Engine - Enterprise
 Engine:
  Version:          19.03.2
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.12.8
  Git commit:       c92ab06
  Built:            Tue Sep  3 15:55:37 2019
  OS/Arch:          linux/amd64
  Experimental:     true
 containerd:
  Version:          1.2.6
  GitCommit:        894b81a4b802e4eb2a91d1ce216b8817763c29fb
 runc:
  Version:          1.0.0-rc8
  GitCommit:        425e105d5a03fabd737a126ad93d62a9eeede87f
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683
```

**Comment by AkihiroSuda** [comment]
buildx docs: `For entitlements to be enabled, the buildkitd daemon also needs to allow them with --allow-insecure-entitlement`

**Comment by bharathappali** [comment]
Thanks @AkihiroSuda . It worked now.

**Comment by uvwild** [comment]
just to add another use case.
I am trying to fix a dockerfile build of an ibmdb2 container with a test database
IBM removed the v10 image from the hub. But the v11 DB image only starts with --privileged.
So all the code setting up the database in the Dockerfile is unfunctional now because the db2 doesn 't start without privileged. :( 
There seem to be a complicated workaround with using docker run and docker commit.
In a productive build pipeline this creates a lot of extra complexity. 

I have to ask like https://github.com/maneamarius  in https://github.com/moby/moby/issues/1916#issuecomment-361173550

Why is it such a big deal to support this? The build does execute a run under the hood. 

In this specific use case a privileged build option would support a kind of "backward compatibility" and I know I am not the only one who had this issue after my web research.

**Comment by bharathappali** [comment]
@uvwild I'm not sure if this helps your use case but you can give a try to build with [kaniko](https://github.com/GoogleContainerTools/kaniko) Your image will be built without a docker deamon and you can extract the image once its done and using kaniko is jus like running a container you can use `--privileged` or `--cap-add <capability which is needed>` which might solve your problem.

I accept its not a complete solution you were expecting but an easier workaround which may fit in your build pipeline.

**Comment by jerblack** [comment]
> buildx docs: `For entitlements to be enabled, the buildkitd daemon also needs to allow them with --allow-insecure-entitlement`

What does this mean? Where is this done? I can't find documentation on this anywhere.

**Comment by dlants** [comment]
I was able to get past the `fuse: device not found, try 'modprobe fuse' first` error (trying to run tup https://gittup.org/tup/ ) by doing the following:

Dockerfile:
```
# syntax=docker/dockerfile:1-labs
...
RUN --security=insecure <command requiring fuse>
```
see: https://docs.docker.com/engine/reference/builder/#run---securityinsecure

commands
```
docker buildx create --buildkitd-flags '--allow-insecure-entitlement security.insecure' --name insecure-builder
docker buildx use insecure-builder
docker buildx build --allow security.insecure .
```
see: https://docs.docker.com/engine/reference/commandline/buildx_build/#allow
see: https://docs.docker.com/engine/reference/commandline/buildx_create/#buildkitd-flags

I don't totally understand all of these settings or if they're wise to use, so beware! Definitely wish there was an easier way to do this!
